src/ShellCheck/Analytics.hs

```diff
@@ -4734,5 +4742,90 @@ checkSetESuppressed params t =
         isIn t cmds = getId t `elem` map getId cmds
 
 
+prop_checkExtraMaskedReturns1  = verifyTree    checkExtraMaskedReturns "read -ra arr <(ls)"
+prop_checkExtraMaskedReturns2  = verifyTree    checkExtraMaskedReturns "ls >(cat)"
+prop_checkExtraMaskedReturns3  = verifyTree    checkExtraMaskedReturns "false | true"
+prop_checkExtraMaskedReturns4  = verifyNotTree checkExtraMaskedReturns "set -o pipefail; false | true"
+prop_checkExtraMaskedReturns5  = verifyNotTree checkExtraMaskedReturns "false | true || true"
+prop_checkExtraMaskedReturns6  = verifyTree    checkExtraMaskedReturns "true $(false)"
+prop_checkExtraMaskedReturns7  = verifyTree    checkExtraMaskedReturns "x=$(false)$(true)"
+prop_checkExtraMaskedReturns8  = verifyTree    checkExtraMaskedReturns "true <<<$(false)"
+prop_checkExtraMaskedReturns9  = verifyNotTree checkExtraMaskedReturns "echo asdf | false"
+prop_checkExtraMaskedReturns10 = verifyNotTree checkExtraMaskedReturns "readonly x=$(false)"
+prop_checkExtraMaskedReturns11 = verifyTree    checkExtraMaskedReturns "readarray -t files < <(ls)"
+checkExtraMaskedReturns params t = runNodeAnalysis checkNode params t
```

Comment:
So far so good! Here are some cases it misses:

False negative:
```
x="$(false)$(false)"      # Check out getWordParts :: Token -> [Token]
x="$(false)""$(false)"    # It takes a T_NormalWord and flattens it into a list of all the quoted and unquoted elements in the string
x=( $(false) $(false) )
x=( $(false) [4]=$(false) )

# These may be less relevant:
cat << foo
 $(false)
foo
x=$(( $(date +%s) ))
[[ $(false) ]]
```

False positive:
```
x=$(false) y=z         # T_SimpleCommand has two assignments in its list in this case
x=$(false) y=$(false)  # Only the first one is masked
```

These are already correctly handled but untested:
```
echo $(( $(date +%s) ))
x=`false``false`
x=( $(false) )
```

Reply:
Thanks for all these test cases. Any suggestion as to how I should approach this?

```bash
x=$(false) y=$(false)  # Only the first one is masked
```

I was thinking that once I find a `T_SimpleCommand`, I'd traverse the tree upwards, looking for a `T_Assignment`. Once I find that, I'd look at the assignments on the the left of the `T_SimpleCommand` and check to see if their descendants contain any `T_SimpleCommand` nodes. The trouble is that it's not obvious to me how I can iterate over the descendants of a node without having a huge `case ... of` statement which essentially handles each rule in Bash's grammar.

Reply:
This big case...of statement already exists. There's a `doAnalysis` which invokes each node in a tree with a monadic action. The obvious monad is Cont, but to avoid importing that, perhaps Maybe:

```
containsSimpleCommand rootnode = isNothing $ doAnalysis (\c -> case c of T_SimpleCommand {} -> fail ""; _ -> return ()) rootnode
```

Currently the code finds the potentially masked T_SimpleCommand, and tries to determine if it is masked. This has some issues like `echo $(t=$(false); echo $t)` emitting two warnings even though only one is actually masked, and `echo $([[ x ]])` emits nothing since it's a compound command instead of a simple command.

An alternative approach is finding the potential masker, e.g. a T_ProcSub, or a `T_SimpleCommand _ [] (cmd:args)` where `any isCommandSubstitution $ concatMap getWordParts args`. Since it seems like this is basically required to recognize the context anyways, maybe this is easier.